include/pybind11/attr.h

```diff
@@ -14,27 +14,22 @@
 
 NAMESPACE_BEGIN(pybind11)
 
-template <typename T> struct arg_t;
-
 /// Annotation for keyword arguments
 struct arg {
     constexpr explicit arg(const char *name) : name(name) { }
 
     template <typename T>
     constexpr arg_t<T> operator=(const T &value) const { return {name, value}; }
-    template <typename T, size_t N>
-    constexpr arg_t<const T *> operator=(T const (&value)[N]) const {
-        return operator=((const T *) value);
-    }
 
     const char *name;
 };
 
 /// Annotation for keyword arguments with default values
 template <typename T> struct arg_t : public arg {
     constexpr arg_t(const char *name, const T &value, const char *descr = nullptr)
-        : arg(name), value(value), descr(descr) { }
-    T value;
+        : arg(name), value(&value), descr(descr) { }
```

Comment:
Right -- that makes sense. Is the restriction to rvalues unavoidable? If it's easy to also support your example
 
``
auto a = "arg"_a=MyArg(1, 2, 3);
f(a);
``

with minor changes then that would be nice. Switching to ``py::object`` for the underlying argument storage in this case is fine to me. The only reason for the current state of affairs is because of the way in which ``def()`` arguments were processed previously. 

Regarding value policies: this unlikely to be a serious concern for default arguments constructed at module load time, and it's always safe to make a copy (which is what will happen for ``const T&`` instances that have not been previously observed by pybind11).